process/sync/storageBootstrap/baseStorageBootstrapper.go

```diff
@@ -188,45 +192,60 @@ func (st *storageBootstrapper) applyBootInfos(bootInfos []bootstrapStorage.Boots
 
 	defer func() {
 		if err != nil {
-			st.blkExecutor.RestoreLastNotarizedHrdsToGenesis()
-			st.forkDetector.RestoreFinalCheckPointToGenesis()
+			st.forkDetector.RestoreToGenesis()
+			st.blockTracker.RestoreToGenesis()
 		}
 	}()
 
 	for i := len(bootInfos) - 1; i >= 0; i-- {
-		log.Debug("apply block",
-			"nonce", bootInfos[i].LastHeader.Nonce,
-			"shardId", bootInfos[i].LastHeader.ShardId)
-
-		lastNotarized := make(map[uint32]*sync.HdrInfo, len(bootInfos[i].LastNotarizedHeaders))
-		for _, lastNotarizedHeader := range bootInfos[i].LastNotarizedHeaders {
-			log.Debug("added notarized header",
-				"nonce", lastNotarizedHeader.Nonce,
-				"shardId", lastNotarizedHeader.ShardId)
-
-			lastNotarized[lastNotarizedHeader.ShardId] = &sync.HdrInfo{
-				Nonce: lastNotarizedHeader.Nonce,
-				Hash:  lastNotarizedHeader.Hash,
-			}
-		}
+		log.Debug("apply header",
+			"shard", bootInfos[i].LastHeader.ShardId,
+			"nonce", bootInfos[i].LastHeader.Nonce)
 
-		err = st.bootstrapper.applyNotarizedBlocks(lastNotarized)
+		err = st.bootstrapper.applyCrossNotarizedHeaders(bootInfos[i].LastCrossNotarizedHeaders)
 		if err != nil {
-			log.Debug("cannot apply notarized block", "error", err.Error())
+			log.Debug("cannot apply cross notarized headers", "error", err.Error())
+			return err
+		}
+
+		selfNotarizedHeadersHashes := make([][]byte, len(bootInfos[i].LastSelfNotarizedHeaders))
+		for index, selfNotarizedHeader := range bootInfos[i].LastSelfNotarizedHeaders {
+			selfNotarizedHeadersHashes[index] = selfNotarizedHeader.Hash
+		}
 
+		selfNotarizedHeaders, err := st.bootstrapper.applySelfNotarizedHeaders(selfNotarizedHeadersHashes)
+		if err != nil {
+			log.Debug("cannot apply self notarized headers", "error", err.Error())
 			return err
 		}
 
-		lastFinalHashes := make([][]byte, 0, len(bootInfos[i].LastFinals))
-		for _, lastFinal := range bootInfos[i].LastFinals {
-			lastFinalHashes = append(lastFinalHashes, lastFinal.Hash)
+		header, err := st.bootstrapper.getHeader(bootInfos[i].LastHeader.Hash)
+		if err != nil {
+			return err
 		}
 
-		err = st.addHeaderToForkDetector(bootInfos[i].LastHeader.Hash, lastFinalHashes)
+		log.Debug("add header to fork detector",
+			"shard", header.GetShardID(),
+			"round", header.GetRound(),
+			"nonce", header.GetNonce(),
+			"hash", bootInfos[i].LastHeader.Hash)
+
+		err = st.forkDetector.AddHeader(header, bootInfos[i].LastHeader.Hash, process.BHProcessed, selfNotarizedHeaders, selfNotarizedHeadersHashes)
 		if err != nil {
-			log.Debug("cannot apply final block", "error", err.Error())
 			return err
 		}
+
+		if i > 0 {
+			log.Debug("added self notarized header in block tracker",
+				"shard", header.GetShardID(),
+				"round", header.GetRound(),
+				"nonce", header.GetNonce(),
+				"hash", bootInfos[i].LastHeader.Hash)
+
+			st.blockTracker.AddSelfNotarizedHeader(st.shardCoordinator.SelfId(), header, bootInfos[i].LastHeader.Hash)
```

Comment:
Because selfNotarizedHeaders / crossNotarizedHeaders lists from block tracker contains only the notarized headers (finals) and not also the headers which attests the finality. Actually it is exactly as we had in the old code from processor. The header which attest the finality of the last notarized would be added in this list later when another one would be constructed above it. So when node bootstrap from storage would apply the same logic as in online.